xla/service/elemental_ir_emitter.cc

```diff
@@ -793,6 +793,123 @@ llvm::Value* EmitF8e4m3b11fnuzToF16(llvm::Value* f8_value,
+    return llvm::ConstantInt::get(b->getInt16Ty(), val);
+  };
+  llvm::Value* f16_bits = b->CreateAnd(
+      b->CreateBitCast(f16_value, b->getInt16Ty()), i16_const(0x7FFF));
+  llvm::Value* is_overflow =
+      b->CreateICmpUGE(f16_bits, i16_const(0x4700));                    // 7.0
+  llvm::Value* is_nan = b->CreateICmpUGT(f16_bits, i16_const(0x7C00));  // inf
+  llvm::Value* max_or_nan =
+      b->CreateSelect(is_nan, i8_const(0x8), i8_const(0x7));
+  llvm::Value* f4_normal_or_overflow =
+      b->CreateSelect(is_overflow, max_or_nan, f4_normal);
+
+  // Special case for exponent underflow.
+  llvm::Value* is_underflow = b->CreateICmpSLE(f4_normal, i8_const(1));
+  llvm::Value* is_one = b->CreateICmpUGE(f16_bits, i16_const(0x3A00));   // 0.75
+  llvm::Value* is_zero = b->CreateICmpULE(f16_bits, i16_const(0x3400));  // 0.25
+  llvm::Value* denorm_or_zero =
+      b->CreateSelect(is_zero, i8_const(0x0), i8_const(0x1));
+  llvm::Value* f4_small =
+      b->CreateSelect(is_one, i8_const(0x2), denorm_or_zero);
+  llvm::Value* f4_result =
+      b->CreateSelect(is_underflow, f4_small, f4_normal_or_overflow);
+
+  // Add sign to the resulting value.
+  return b->CreateOr(f4_result, b->CreateShl(f4_sign, 3));
+}
+
+llvm::Value* EmitF4e2m1fnToF16(llvm::Value* f8_value, llvm::IRBuilder<>* b) {
+  llvm::Value* as_int16 = b->CreateZExt(f8_value, b->getInt16Ty());
+
+  // Extract sign, exponent and mantissa from reduced precision value.
+  auto i16_const = [&](int val) {
+    return llvm::ConstantInt::get(b->getInt16Ty(), val);
+  };
+  llvm::Value* sign = b->CreateLShr(as_int16, 3);
+  llvm::Value* sign_shifted = b->CreateShl(sign, 15);
+  llvm::Value* bits = b->CreateAnd(as_int16, i16_const(0x7));
+  llvm::Value* bits_shifted = b->CreateShl(bits, 9);
+
+  // Re-bias the exponent and handle denormals.
+  llvm::Value* f16_normal = b->CreateAdd(bits_shifted, i16_const(14 << 10));
+  llvm::Value* is_denorm_or_zero = b->CreateICmpULE(bits, i16_const(1));
+  llvm::Value* is_zero = b->CreateICmpEQ(bits, i16_const(0));
+  llvm::Value* denorm_or_zero =
+      b->CreateSelect(is_zero, i16_const(0x0000), i16_const(0x3800));
+  llvm::Value* f16_result =
+      b->CreateSelect(is_denorm_or_zero, denorm_or_zero, f16_normal);
+
+  // Add sign to the resulting value.
+  llvm::Value* f16_signed = b->CreateOr(f16_result, sign_shifted);
+  return b->CreateBitCast(f16_signed, b->getHalfTy());
+}
+
+llvm::Value* EmitF32ToF8e8m0fnu(llvm::Value* f32_value, llvm::IRBuilder<>* b) {
+  llvm::Value* as_int32 = b->CreateBitCast(f32_value, b->getInt32Ty());
+
+  // Result is NaN if input is zero, negative, infinity or NaN.
+  auto i32_const = [&](int val) {
+    return llvm::ConstantInt::get(b->getInt32Ty(), val);
+  };
+  llvm::Value* is_denorm = b->CreateICmpULE(as_int32, i32_const(0x400000));
+  llvm::Value* is_nan =
+      b->CreateOr(b->CreateICmpSLE(as_int32, i32_const(0)),
+                  b->CreateICmpSGE(as_int32, i32_const(0x7F400000)));
```

Comment:
Comment what 0x7F400000 is, e.g. `// 1.5 * 2^127`. And same for 0x400000. And similarly in EmitF8e8m0fnuToF32.